contracts/exchangeIssuance/ExchangeIssuance.sol

```diff
@@ -0,0 +1,492 @@
+            uint256 unit = uint256(positions[i].unit);
+            address token = positions[i].component;
+            uint256 amountOut = minSetTokenAmountOut.mul(unit).div(1 ether);
+            uint256 uniPrice = _tokenAvailable(uniFactory, token) ? _getBuyPrice(true, token, amountOut) : PreciseUnitMath.maxUint256();
+            uint256 sushiPrice = _tokenAvailable(sushiFactory, token) ? _getBuyPrice(false, token, amountOut) : PreciseUnitMath.maxUint256();
+            uint256 amountEth = Math.min(uniPrice, sushiPrice);
+            sumEth = sumEth.add(amountEth);
+            amountEthIn[i] = amountEth;
+        }
+
+        uint256 maxIndexAmount = PreciseUnitMath.maxUint256();
+        for (uint i = 0; i < positions.length; i++) {
+            address token = positions[i].component;
+            uint256 unit = uint256(positions[i].unit);
+
+            uint256 scaledAmountEth = amountEthIn[i].mul(wethBalance).div(sumEth);  // scale the amountEthIn
+            uint256 uniTokenOut = _tokenAvailable(uniFactory, token) ? _getBuyPriceExactETH(true, token, scaledAmountEth) : 0;
+            uint256 sushiTokenOut = _tokenAvailable(sushiFactory, token) ? _getBuyPriceExactETH(false, token, scaledAmountEth) : 0;
+
+            uint256 amountTokenOut = Math.max(uniTokenOut, sushiTokenOut);
+            // update the maxIndexAmount
+            maxIndexAmount = Math.min(amountTokenOut.mul(1 ether).div(unit), maxIndexAmount);
+        }
+        return maxIndexAmount;
+    }
+
+    /* ============ Internal Functions ============ */
+
+    /**
+     * Sells the total balance that the contract holds of each component of the set
+     * using the best quoted price from either Uniswap or Sushiswap
+     * 
+     * @param _setToken     The set token that is being liquidated
+     */
+    function _liquidateComponents(ISetToken _setToken) internal {
+        ISetToken.Position[] memory positions = _setToken.getPositions();
+        for (uint256 i=0; i<positions.length; i++) {
+            _sellTokenBestPrice(positions[i].component);
+        }
+    }
+
+    /**
+     * Handles converting the contract's full WETH balance to the output
+     * token or ether and transfers it to the msg sender.
+     *
+     * @param _isOutputETH      Converts the contract's WETH balance to ETH if set to true
+     * @param _outputToken      The token to swap the contract's WETH balance to. 
+     *                          Ignored if _isOutputETH is set to true.
+     * @param _minOutputReceive Minimum amount of output token or ether to receive. This 
+     *                          function reverts if the output is less than this.
+     */
+    function _handleRedeemOutput(bool _isOutputETH, address _outputToken, uint256 _minOutputReceive) internal {
+        if(_isOutputETH) {
+            IWETH(WETH).withdraw(IERC20(WETH).balanceOf(address(this)));
+            require(address(this).balance > _minOutputReceive, "INSUFFICIENT_OUTPUT_AMOUNT");
```

Comment:
The error message can indicate the name of the contract as well. Something like,
`ExchangeIssuance: INSUFFICIENT_OUTPUT_AMOUNT`. Same applies for other error messages as well.